pkg/querier/querier.go

```diff
@@ -373,11 +377,96 @@ func (q querier) Select(_ bool, sp *storage.SelectHints, matchers ...*labels.Mat
 
 // LabelsValue implements storage.Querier.
 func (q querier) LabelValues(name string) ([]string, storage.Warnings, error) {
-	return q.metadataQuerier.LabelValues(name)
+	if !q.queryStoreForLabels {
+		return q.metadataQuerier.LabelValues(name)
+	}
+
+	if len(q.queriers) == 1 {
+		return q.queriers[0].LabelValues(name)
+	}
+
+	// Using an errgroup here instead of channels, etc because this
+	// is a better model imo and we should move to this everywhere.
+	var (
+		g, _     = errgroup.WithContext(q.ctx)
+		sets     = [][]string{}
+		warnings = storage.Warnings(nil)
+
+		resMtx sync.Mutex
+	)
+
+	for _, querier := range q.queriers {
+		// Need to reassign as the original variable will change and can't be relied on in a goroutine.
+		querier := querier
+		g.Go(func() error {
+			myValues, myWarnings, err := querier.LabelValues(name)
+			if err != nil {
+				return err
+			}
+
+			// We need values to be sorted we can merge them.
+			sort.Strings(myValues)
```

Comment:
May you confirm store-gateway and "distributor" don't return it already sorted, right? If store-gateway does, then we could add it to the API spec (response is guaranteed to be sorted) and we could just sort it in the "distributor".

_Same comment applies to `LabelNames()`_

Reply:
I have this change made in the distributor, but but the querier interface we rely on is the Prometheus interface which doesn't guarantee sorted-ness.

We can always say in Cortex we always return sorted, but not sure if it will bite us sometime in the future.